cellxgene_schema_cli/tests/test_utils.py

```diff
@@ -71,3 +79,31 @@ def test_map_ontology_term__(adata_without_raw):
     assert all(a == "CL:0000001" for a in donor_1_rows["cell_type_ontology_term_id"])
     donor_2_rows = adata_without_raw.obs.loc[adata_without_raw.obs["donor_id"] == "donor_2"]
     assert all(a == "CL:0000002" for a in donor_2_rows["cell_type_ontology_term_id"])
+
+
+@pytest.fixture
+def noncanonical_matrix():
+    array = np.array([[1, 0, 1], [3, 2, 3], [4, 5, 4]])
+    return coo_matrix((array[0], (array[1], array[2])))
+
+
+class TestEnforceCanonical:
+    def test_adata_with_noncanonical_X_and_raw_X(self, noncanonical_matrix):
+        assert noncanonical_matrix.has_canonical_format is False
+        adata = AnnData(noncanonical_matrix)
+        enforce_canonical_format(adata)
+        assert adata.X.has_canonical_format is True
+
+    def test_adata_with_noncanonical_raw_X(self, noncanonical_matrix):
+        assert noncanonical_matrix.has_canonical_format is False
+        adata = AnnData(raw=AnnData(noncanonical_matrix))
```

Comment:
interesting, you can pass an entire AnnData as `raw` param? Same as `raw=AnnData(noncanonical_matrix).X`?

Reply:
you can 😄  https://anndata.readthedocs.io/en/0.7.8/generated/anndata.AnnData.raw.html#anndata.AnnData.raw